File: dotnet/src/Pagou.Examples.Core/Reconcile.cs
namespace Pagou.Examples.Core;

public enum FulfillmentDecision
{
    Fulfill,
    Wait,
    Cancel,
}

/// <summary>
/// Server-side reconciliation: fetch the transaction from the API (the source of
/// truth) and decide whether to fulfill. Business state changes only on this
/// confirmed result, never on an unverified webhook body.
/// </summary>
public static class Reconciler
{
    /// <summary>Statuses at which a charge is settled and it is safe to fulfill.</summary>
    private static readonly HashSet<string> TerminalPaidStatuses = new() { "paid", "captured" };

    /// <summary>Terminal failure/cancel states: stop waiting, release the order.</summary>
    private static readonly HashSet<string> TerminalFailedStatuses = new() { "canceled", "expired", "refused" };

    /// <summary>Maps a transaction status to a business decision. Never fulfill on a pending state.</summary>
    public static FulfillmentDecision DecideFulfillment(string status)
    {
        if (TerminalPaidStatuses.Contains(status))
        {
            return FulfillmentDecision.Fulfill;
        }
        if (TerminalFailedStatuses.Contains(status))
        {
            return FulfillmentDecision.Cancel;
        }
        return FulfillmentDecision.Wait;
    }

    public static async Task<(Transaction Transaction, FulfillmentDecision Decision)?> ReconcileTransactionAsync(
        string id, PagouHttpClient client, CancellationToken ct = default)
    {
        try
        {
            var result = await client.RequestDataAsync<Transaction>(
                new RequestParams { Method = HttpMethod.Get, Path = $"/v2/transactions/{id}" }, ct)
                .ConfigureAwait(false);
            return (result.Data, DecideFulfillment(result.Data.Status));
        }
        catch (NotFoundException)
        {
            return null;
        }
    }
}

/// <summary>Statuses from which a transfer can typically be cancelled.</summary>
public static class TransferStatuses
{
    public static readonly HashSet<string> Cancelable = new() { "pending", "scheduled" };
}

File: php/src/Reconcile.php
<?php

declare(strict_types=1);

namespace Pagou\Examples;

final class Reconcile
{
    public const FULFILL = 'fulfill';
    public const WAIT = 'wait';
    public const CANCEL = 'cancel';

    /** Maps a transaction status to a business decision. Never fulfill on a pending state. */
    public static function decideFulfillment(string $status): string
    {
        if (in_array($status, Types::TERMINAL_PAID_STATUSES, true)) {
            return self::FULFILL;
        }
        if (in_array($status, Types::TERMINAL_FAILED_STATUSES, true)) {
            return self::CANCEL;
        }
        return self::WAIT;
    }

    /**
     * Server-side reconciliation: fetch the transaction from the API (the source
     * of truth) and decide whether to fulfill. Business state changes only on
     * this confirmed result, never on an unverified webhook body.
     *
     * @return array{transaction: array<string, mixed>, decision: string}|null
     */
    public static function transaction(string $id, ?HttpClient $client = null): ?array
    {
        $client ??= new HttpClient();
        try {
            $result = $client->requestData(['method' => 'GET', 'path' => "/v2/transactions/{$id}"]);
        } catch (NotFoundError) {
            return null;
        }

        $transaction = is_array($result['data']) ? $result['data'] : [];
        return [
            'transaction' => $transaction,
            'decision' => self::decideFulfillment((string) ($transaction['status'] ?? '')),
        ];
    }
}
